ext/quic/lib.rs

```diff
@@ -0,0 +1,606 @@
+  ops = [
+    op_quic_listen<P>,
+    op_quic_accept,
+    op_quic_connect<P>,
+    op_quic_accept_bi,
+    op_quic_accept_uni,
+    op_quic_open_bi,
+    op_quic_open_uni,
+    op_quic_max_datagram_size,
+    op_quic_send_datagram,
+    op_quic_read_datagram,
+    op_quic_close_connection,
+    op_quic_close_endpoint,
+    op_quic_connection_closed,
+    op_quic_get_send_stream_priority,
+    op_quic_set_send_stream_priority,
+    op_quic_get_conn_remote_addr,
+  ],
+  esm = ["01_quic.js"],
+  options = {
+    root_cert_store_provider: Option<Arc<dyn RootCertStoreProvider>>,
+    unsafely_ignore_certificate_errors: Option<Vec<String>>,
+  },
+  state = |state, options| {
+    state.put(DefaultTlsOptions {
+      root_cert_store_provider: options.root_cert_store_provider,
+    });
+    state.put(UnsafelyIgnoreCertificateErrors(
+      options.unsafely_ignore_certificate_errors,
+    ));
+  },
+);
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Addr {
+  pub hostname: String,
+  pub port: u16,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ListenArgs {
+  cert: String,
+  key: String,
+  alpn_protocols: Option<Vec<String>>,
+}
+
+struct EndpointResource(quinn::Endpoint);
+
+impl Resource for EndpointResource {
+  fn name(&self) -> Cow<str> {
+    "quicListener".into()
+  }
+}
+
+#[op2(async)]
+#[serde]
+async fn op_quic_listen<NP>(
+  state: Rc<RefCell<OpState>>,
+  #[serde] addr: Addr,
+  #[serde] args: ListenArgs,
+) -> Result<(ResourceId, Addr), AnyError>
+where
+  NP: NetPermissions + 'static,
+{
+  state
+    .borrow_mut()
+    .borrow_mut::<NP>()
+    .check_net(&(&addr.hostname, Some(addr.port)), "Deno.listenQuic()")?;
+
+  let cert_chain = load_certs(&mut BufReader::new(args.cert.as_bytes()))?;
+  let key_der = load_private_keys(args.key.as_bytes())?.remove(0);
+
+  let addr = resolve_addr(&addr.hostname, addr.port)
+    .await?
+    .next()
+    .ok_or_else(|| generic_error("No resolved address found"))?;
+
+  let mut crypto = rustls::ServerConfig::builder()
+    .with_safe_defaults()
+    .with_no_client_auth()
+    .with_single_cert(cert_chain, key_der)?;
+  if let Some(alpn_protocols) = args.alpn_protocols {
+    crypto.alpn_protocols = alpn_protocols
+      .into_iter()
+      .map(|alpn| alpn.into_bytes())
+      .collect();
+  }
+  let config = quinn::ServerConfig::with_crypto(Arc::new(crypto));
+  let endpoint = quinn::Endpoint::server(config, addr)?;
+
+  let addr = endpoint.local_addr()?;
+  let addr = Addr {
+    hostname: format!("{}", addr.ip()),
+    port: addr.port(),
+  };
+
+  let rid = state
+    .borrow_mut()
+    .resource_table
+    .add(EndpointResource(endpoint));
+  Ok((rid, addr))
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CloseInfo {
+  close_code: u64,
+  reason: String,
+}
+
+#[op2]
+fn op_quic_close_endpoint(
+  state: Rc<RefCell<OpState>>,
+  #[smi] rid: ResourceId,
+  #[serde] info: CloseInfo,
```

Comment:
I would suggest to use two separate arguments here - accepting `#[smi]` and `#[string]` it gonna be way faster than a `#[serde]` argument. For reference here's a nice overview doc with all the possible types that `op2` accepts:
https://github.com/denoland/deno_core/tree/main/ops/op2#parameters